code/_hooks/events.dm

```diff
@@ -51,3 +51,133 @@
 		args["event"] = src
 		if(call(objRef,procName)(args, holder)) //An intercept value so whatever code section knows we mean business
 			. = 1
+
+#define EVENT_HANDLER_OBJREF_INDEX 1
+#define EVENT_HANDLER_PROCNAME_INDEX 2
+
+/proc/CallAsync(datum/source, proctype, list/arguments)
+	set waitfor = FALSE
+	return call(source, proctype)(arglist(arguments))
+
+// Declare children of this type path to use as identifiers for the events.
+/lazy_event
+
+// TODO: Document here the arguments that need to be passed to the procs invoked by each event
+/lazy_event/on_z_transition
+/lazy_event/on_post_z_transition
+
+/datum
+	/// Associative list of type path -> list(),
+	/// where the type path is a descendant of /event_type.
+	/// The inner list is itself an associative list of string -> list(),
+	/// where string is the \ref of an object + the proc to be called.
+	/// The list associated with the string above contains the hard-ref
+	/// to an object and the proc to be called.
+	var/list/list/registered_events
+
+/**
+  * Calls all registered event handlers with the specified parameters, if any.
+  * Arguments:
+  * * lazy_event/event_type Required. The typepath of the event to invoke.
+  * * list/arguments Optional. List of parameters to be passed to the event handlers.
+  */
+/datum/proc/lazy_invoke_event(lazy_event/event_type, list/arguments)
+	SHOULD_NOT_OVERRIDE(TRUE)
+	if(!length(registered_events))
+		// No event at all is registered for this datum.
+		return
+	var/list/event_handlers = registered_events[event_type]
+	if(!length(event_handlers))
+		// This datum does not have any handler registered for this event_type.
+		return
+	. = NONE
+	for(var/key in event_handlers)
+		var/list/handler = event_handlers[key]
+		var/objRef = handler[EVENT_HANDLER_OBJREF_INDEX]
+		var/procName = handler[EVENT_HANDLER_PROCNAME_INDEX]
+		. |= CallAsync(objRef, procName, arguments)
```

Comment:
Why is it `|=` and not `+=` ?

Reply:
Why not? `|` lets you return bitflags, `+` would be less useful, I think. If the need for a different way of handling return values arises we'll change it but at the moment this is unused.

Reply:
Fair enough. Not like it would matter a lot in that context.